subsys/bluetooth/host/smp.c

```diff
@@ -4369,8 +4374,8 @@ static int bt_smp_recv(struct bt_l2cap_chan *chan, struct net_buf *buf)
 
 	if (!atomic_test_and_clear_bit(smp->allowed_cmds, hdr->code)) {
 		LOG_WRN("Unexpected SMP code 0x%02x", hdr->code);
-		/* Don't send error responses to error PDUs */
-		if (hdr->code != BT_SMP_CMD_PAIRING_FAIL) {
+		/* Do not send errors outside of pairing procedure. */
+		if (is_in_pairing_procedure(smp)) {
```

Comment:
Doesn't this change the behaviour when receiving a Pairing Failed PDU during a pairing procedure (when the flag is set)? I.e. previously the code would not send a response to it, but now it will?

Reply:
No it will not send a response because Pairing Failed is an allowed command during the pairing procedure.

Reply:
Ah, so you mean it's covered by the `if (!atomic_test_and_clear_bit(smp->allowed_cmds, hdr->code))` check, meaning we never get into this branch in that case, meaning this internal branch was previously dead code?

Reply:
Yes, it's covered by `if (!atomic_test_and_clear_bit(smp->allowed_cmds, hdr->code))`. But no, it was not dead code previously. When `smp_error` is called it send a `BT_SMP_CMD_PAIRING_FAIL` PDU. 

So, previously if we received a SMP PDU that was not an allowed command, we were sending back an error with a Pairing Failed PDU. To avoid a infinite loop of Pairing Failed exchange I guess, we were not sending it if the not allowed command was a Pairing Failed PDU.

Back to my update, we now don't send back Pairing Failed PDUs when we receive not allowed commands *except* if we are in the pairing procedure. Then we still use `smp_error` which will send a Pairing Failed PDUs.